### PR TITLE
admission: incorporate L0 size into overload score

### LINKED ISSUE #109207: admission: discount sub-levels when L0 bytes are small
We have seen an example (in a clearrange roachtest) where small ingests (likely ranges that have shrunk and being merged) happen at a high rate (since they are small, not many bytes needed to be transferred to the node), resulting in 6+ flushes/s (since these ingests overlap with the memtable). This increases the sub-level count quickly, causing admission control to throttle user-facing work.

The following is an example of a sequence of such ingests, which started with an empty L0, and caused sub-level count to increase to 15+. We do want to use the latest state of L0, since smoothing would delay the reaction to overload, but in this case there isn't a real overload since these bytes can be compacted out of L0 is a few seconds. We should discount the sub-level count based on the bytes in L0 to prevent overreaction in admission control.

```
I230817 20:53:30.010702 10124640 3@pebble/event.go:709 ⋮ [n2,s2,pebble] 10912  [JOB 3027] ingested as flushable 027239 (1.3KB), 027244 (1.0KB), 027240 (1.7KB), 027241 (1.1KB), 027242 (1.0KB), 027243 (1.0KB)
I230817 20:53:30.010818 10124955 3@pebble/event.go:681 ⋮ [n2,s2,pebble] 10913  [JOB 3030] flushing 1 memtable (1.2MB) to L0
I230817 20:53:30.020320 10124955 3@pebble/event.go:685 ⋮ [n2,s2,pebble] 10916  [JOB 3030] flushed 1 memtable (1.2MB) to L0 [027247] (214KB), in 0.0s (0.0s total), output rate 23MB/s
I230817 20:53:30.023264 10124908 3@pebble/event.go:685 ⋮ [n2,s2,pebble] 10919  [JOB 3031] flushed 6 ingested flushables L0:027239 (1.3KB) + L0:027244 (1.0KB) + L0:027240 (1.7KB) + L0:027241 (1.1KB) + L0:027242 (1.0KB) + L6:027243 (1.0KB) in 0.0s (0.0s total), output rate 10MB/s
I230817 20:53:30.315841 10126769 3@pebble/event.go:709 ⋮ [n2,s2,pebble] 10933  [JOB 3034] ingested as flushable 027255 (1.3KB), 027260 (1.0KB), 027256 (1.7KB), 027257 (1.1KB), 027258 (1.0KB), 027259 (1.0KB)
I230817 20:53:30.315888 10127180 3@pebble/event.go:681 ⋮ [n2,s2,pebble] 10934  [JOB 3037] flushing 1 memtable (1.3MB) to L0
I230817 20:53:30.325277 10127180 3@pebble/event.go:685 ⋮ [n2,s2,pebble] 10936  [JOB 3037] flushed 1 memtable (1.3MB) to L0 [027263] (192KB), in 0.0s (0.0s total), output rate 26MB/s
I230817 20:53:30.327468 10127273 3@pebble/event.go:685 ⋮ [n2,s2,pebble] 10938  [JOB 3038] flushed 6 ingested flushables L0:027255 (1.3KB) + L0:027260 (1.0KB) + L0:027256 (1.7KB) + L0:027257 (1.1KB) + L0:027258 (1.0KB) + L6:027259 (1.0KB) in 0.0s (0.0s total), output rate 6.3MB/s
I230817 20:53:30.466058 10128340 3@pebble/event.go:709 ⋮ [n2,s2,pebble] 10949  [JOB 3041] ingested as flushable 027266 (1.3KB), 027271 (1.0KB), 027267 (1.6KB), 027268 (1.1KB), 027269 (1.0KB), 027270 (1.0KB)
I230817 20:53:30.467942 10128498 3@pebble/event.go:681 ⋮ [n2,s2,pebble] 10950  [JOB 3044] flushing 1 memtable (625KB) to L0
I230817 20:53:30.481680 10128498 3@pebble/event.go:685 ⋮ [n2,s2,pebble] 10952  [JOB 3044] flushed 1 memtable (625KB) to L0 [027274] (109KB), in 0.0s (0.0s total), output rate 8.3MB/s
I230817 20:53:30.485525 10128343 3@pebble/event.go:685 ⋮ [n2,s2,pebble] 10954  [JOB 3046] flushed 6 ingested flushables L0:027266 (1.3KB) + L0:027271 (1.0KB) + L0:027267 (1.6KB) + L0:027268 (1.1KB) + L0:027269 (1.0KB) + L0:027270 (1.0KB) in 0.0s (0.0s total), output rate 9.7MB/s
I230817 20:53:30.902125 10124410 3@pebble/compaction_picker.go:1301 ⋮ [n2,s2,pebble] 10959  pickAuto: L0->L2
I230817 20:53:30.902125 10124410 3@pebble/compaction_picker.go:1301 ⋮ [n2,s2,pebble] 10959 +‹  *L0: 400.0    0.0    0.0     45MB     8.0EB›
I230817 20:53:30.902125 10124410 3@pebble/compaction_picker.go:1301 ⋮ [n2,s2,pebble] 10959 +‹   L2:   0.0    0.0    0.0       0B      64MB›
I230817 20:53:30.902125 10124410 3@pebble/compaction_picker.go:1301 ⋮ [n2,s2,pebble] 10959 +‹   L3:   0.1    0.1    0.1     31MB     461MB›
I230817 20:53:30.902125 10124410 3@pebble/compaction_picker.go:1301 ⋮ [n2,s2,pebble] 10959 +‹   L4:   3.0    3.0    0.9    136GB     3.2GB›
I230817 20:53:30.902125 10124410 3@pebble/compaction_picker.go:1301 ⋮ [n2,s2,pebble] 10959 +‹   L5:   3.0    2.9    1.0     68GB      23GB  [L5->L6]›
I230817 20:53:30.902125 10124410 3@pebble/compaction_picker.go:1301 ⋮ [n2,s2,pebble] 10959 +‹   L6:   1.0    1.0    1.0    163GB     168GB›
```

compactions
```
I230817 20:53:30.939837 10131663 3@pebble/event.go:673 ⋮ [n2,s2,pebble] 11016  [JOB 3050] compacted(default) L0 [027247 027239 027244 027240 027241 027242 027263 027255 027260 027256 027257 027258 027274] (527KB) + L2 [] (0B) -> L2 [027278] (365KB), in 0.0s (0.0s total), output rate 10MB/s

I230817 20:53:30.959691 10131981 3@pebble/event.go:673 ⋮ [n2,s2,pebble] 11033  [JOB 3052] compacted(default) L0 [027266 027271 027267 027268 027269 027270] (7.0KB) + L2 [027278] (365KB) -> L2 [027279] (366KB), in 0.0s (0.0s total), output rate 21MB/s
```

@bananabrick 

Jira issue: CRDB-30822

### PR DESCRIPTION
**Pull request by sumeerbhola**
Small ingests happening at a high rate can elevate the sub-level count when small memtables need to be flushed (as observed in a clearrange roachtest). This causes admission control to restrict user-facing traffic, even when the bytes in L0 are modest and can all be compacted away very quickly.

This PR updates the scoring logic to add a max and min sub-level bound. The max is the L0 bytes divided by admission.l0_min_size_per_sub_level, where the latter defaults to 5MB. The min is the sub-level count divided by 3.

Fixes #109207

Epic: none

Release note (ops change): Cluster setting
admission.l0_min_size_per_sub_level is introduced, which reduces the probability of admission control throttling when there is a sequence of small memtable flushes or tiny files ingested into L0.

### CODE REVIEW

File: pkg/util/admission/admissionpb/io_threshold_test.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestIOThreshold(t *testing.T) {
+	var testCases = []struct {
+		iot   IOThreshold
```
Comment (irfansharif) [inline pkg/util/admission/admissionpb/io_threshold_test.go:22]: ([#iot](https://cockroachlabs.slack.com/archives/C05NBPFC7KL/p1692305159757809) has a different connotation these days 😛)

File: pkg/util/admission/admissionpb/io_threshold.go [hunk 2]
```diff
@@ -42,9 +41,26 @@ func (iot *IOThreshold) Score() (float64, bool) {
...
 	}
+	numSubLevels := iot.L0NumSubLevels
+	if iot.L0MinimumSizePerSubLevel > 0 {
+		// Upper-bound on number of sub-levels. See the comment for the cluster
+		// setting admission.l0_sub_level_count_overload_threshold.
+		maxNumSubLevels := int64(math.Round(float64(iot.L0Size) / float64(iot.L0MinimumSizePerSubLevel)))
+		// Say numSubLevels is 30 and maxNumSubLevels is 5. We don't want to
+		// ignore the huge disparity, since that could allow the numSubLevels to
+		// grow to a huge value. So we place a lower-bound. The divisor of 3 was
+		// chosen somewhat arbitrarily.
```
Comment (irfansharif) [inline pkg/util/admission/admissionpb/io_threshold.go:52]: Perhaps add a sentence here about how the max can actually be lower than the min but the min takes precedence.

File: pkg/util/admission/io_load_listener.go [hunk 3]
```diff
@@ -84,6 +84,28 @@ var L0SubLevelCountOverloadThreshold = settings.RegisterIntSetting(
...
+// sstables that overlap with the memtable, and each flush may generate a
+// sub-level. We have seen situations where these flushes have a tiny amount
+// of bytes, but a sequence of these can result in a high sub-level count.
+// The default of 5MB is chosen since:
+//
+//   - 5MB*l0SubLevelCountOverloadThreshold=100MB, which can be very quickly
+//     compacted into Lbase (say Lbase overlapping bytes are 200MB, this is a
+//     100MB+200MB=300MB compaction, which takes < 15s).
+//
+//   - it is significantly smaller than a 64MB memtable, even after accounting
```
Comment (irfansharif) [inline pkg/util/admission/io_load_listener.go:99]: Could you explain for me why this matters?

File: pkg/util/admission/testdata/io_load_listener [hunk 4]
```diff
@@ -650,3 +650,47 @@ compaction score 0.250 (10 ssts, 5 sub-levels), L0 growth 0 B (write 0 B ingest
...
+store-request-estimates: writeTokens: 1
+tick: 0, setAdmittedDoneModelsLocked: l0-write-lm: 1.75x+1 l0-ingest-lm: 0.75x+1 ingest-lm: 1.00x+1
+setAvailableTokens: io-tokens=1105 elastic-disk-bw-tokens=unlimited max-byte-tokens=1105 max-disk-bw-tokens=unlimited
+
+# Score is equal to max sub-levels which is 50000/5000=10.
```
Comment (irfansharif) [inline pkg/util/admission/testdata/io_load_listener:689]: Nit: Say something like "score is capped by max sub-levels, so (50000/5000)/20".

**Review by irfansharif — APPROVED** [review]

**Review by bananabrick — COMMENTED** [review]
Logic looks fine. Does this need a run on clearrange to ensure that it fixes the problem observed in https://github.com/cockroachdb/cockroach/issues/109207.

> Reviewed 4 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109332#-:-NcdSfR53N6-3jA7rAIW:b-ozjo02)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @sumeerbhola)

___
*[`pkg/util/admission/io_load_listener.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109332#gh-1304621114:-NcdRNazEkau0rj9O7hx:b-m8c5vr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c971a44a722a6a0b106fe5a0e5029ed38ab0a2/pkg/util/admission/io_load_listener.go#L99)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Could you explain for me why this matters?
</blockquote></details>

Does this justify the fact that this new sublevel logic shouldn't kick in for the general case where we don't flush tiny memtables?

**Review by sumeerbhola — COMMENTED** [review]
> Does this need a run on clearrange to ensure that it fixes the problem

Yes, worth doing. Which clearrange roachtest was the one you ran, whose logs I looked at?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109332#-:-NcdXhSzDK405G0abwcP:b-8ic95)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick and @irfansharif)

___
*[`pkg/util/admission/io_load_listener.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109332#gh-1304621114:-NcdMSVM7BVTk2vsqLjF:b-tbhmjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c971a44a722a6a0b106fe5a0e5029ed38ab0a2/pkg/util/admission/io_load_listener.go#L99)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Could you explain for me why this matters?
</blockquote></details>

I've rephrased this to not be the second bullet and to say

```
// NB: 5MB is typically significantly smaller than the flush size of a 64MB  
// memtable (after accounting for compression when flushing the memtable). If  
// it were comparable, this size based computation of sub-levels would  
// typically override the actual sub-levels, which would defeat the point of  
// using the sub-level count as a metric to guide admission.
```
___
*[`pkg/util/admission/admissionpb/io_threshold.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109332#gh-1304619906:-NcdMyIiEXeRtWSZ-Izk:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c971a44a722a6a0b106fe5a0e5029ed38ab0a2/pkg/util/admission/admissionpb/io_threshold.go#L52)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Perhaps add a sentence here about how the max can actually be lower than the min but the min takes precedence.
</blockquote></details>

Done
___
*[`pkg/util/admission/testdata/io_load_listener` line 689 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109332#gh-1304624175:-NcdNKdu6Vi3t1OEPtjr:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c971a44a722a6a0b106fe5a0e5029ed38ab0a2/pkg/util/admission/testdata/io_load_listener#L689)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Nit: Say something like "score is capped by max sub-levels, so (50000/5000)/20".
</blockquote></details>

Done
___
*[`pkg/util/admission/admissionpb/io_threshold_test.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109332#gh-1304617846:-NcdNMD3BtXi6E_YEgTl:b32vhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c971a44a722a6a0b106fe5a0e5029ed38ab0a2/pkg/util/admission/admissionpb/io_threshold_test.go#L22)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

([#iot](https://cockroachlabs.slack.com/archives/C05NBPFC7KL/p1692305159757809) has a different connotation these days 😛)
</blockquote></details>

:)

**Review by bananabrick — COMMENTED** [review]
clearrange/checks=true

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109332#-:-Nce0hBp4S768-MVCxG0:bxnjcee)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)

**Comment by sumeerbhola** [comment]
bors r=irfansharif,bananabrick

**Pull request merged**